src/mapping/query/functions.rs

```diff
@@ -210,6 +211,161 @@ impl Function for ToTimestampFn {
 
 //------------------------------------------------------------------------------
 
+#[derive(Debug)]
+pub(in crate::mapping) struct UpcaseFn {
+    query: Box<dyn Function>,
+}
+
+impl UpcaseFn {
+    pub(in crate::mapping) fn new(query: Box<dyn Function>) -> Self {
+        Self { query }
+    }
+}
+
+impl Function for UpcaseFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        match self.query.execute(ctx)? {
+            Value::Bytes(bytes) => Ok(Value::Bytes(
+                bytes.iter().map(|c| c.to_ascii_uppercase()).collect(),
```

Comment:
This isn't going to cope with unicode strings. Would it not be better to do something like:

```rust
std::str::from_utf8(&bytes).unwrap().chars().map(|c| c.to_uppercase())
```

(obviously with handling the unwrap properly)